src/front/glsl/parser/declarations.rs

```diff
@@ -199,8 +199,25 @@ impl<'source> ParsingContext<'source> {
             // TODO: Should we try to make constants here?
             // This is mostly a hack because we don't yet support adding
             // bodies to entry points for variable initialization
-            let maybe_constant =
-                init.and_then(|(root, meta)| parser.solve_constant(ctx.ctx, root, meta).ok());
+
+            // Initializers for external declarations must be constants,
+            // so if `solve_constant` failed, that's an error the user should see.
+            // But if it's an initializer for a local variable, then failing to
+            // produce a constant just means we emit the expression as code;
+            // that is not an error.
+            let maybe_constant = if let Some((root, meta)) = init {
+                let expr = &ctx.ctx.expressions[root];
+                match *expr {
+                    Expression::Load { .. } => None,
+                    _ => match parser.solve_constant(ctx.ctx, root, meta) {
+                        Ok(res) => Some(res),
+                        Err(err) if ctx.external => return Err(err),
+                        _ => None,
+                    },
+                }
```

Comment:
I don't know why but github didn't send my comment which said this wasn't the right way to check.

You need to check the declaration qualifiers not the type of expression, see `DeclarationContext::qualifiers`